constraints/src/main/java/gov/nasa/jpl/aerie/constraints/time/Spans.java

```diff
@@ -64,6 +69,37 @@ public Spans filter(final Predicate<Window> filter) {
     return new Spans(this.windows.stream().filter(filter).toList());
   }
 
+  @Override
+  public Spans split(final int numberOfSubSpans) {
+    if (numberOfSubSpans == 1) {
+      return new Spans(this);
+    }
+    return this.flatMap(x -> {
+      // Width of each sub-window, rounded down to the microsecond
+      final var width = Duration.divide(Duration.subtract(x.end, x.start), numberOfSubSpans);
+
+      final var numberOfMicroSeconds = Duration.subtract(x.end, x.start).in(Duration.MICROSECOND);
+
+      // We throw an exception if the interval contains fewer microseconds than the requested number of sub-spans.
+      if (x.isSingleton()) {
+        throw new UnsplittableIntervalException("Cannot split an instantaneous interval into " + numberOfSubSpans + " pieces.");
+      } else if (numberOfMicroSeconds < numberOfSubSpans) {
+        throw new UnsplittableIntervalException("Cannot split an interval only " + numberOfMicroSeconds + " microseconds long into " + numberOfSubSpans + " pieces.");
+      }
+
+      var cursor = Duration.add(x.start, width);
+      final List<Window> ret = new ArrayList<>();
+      ret.add(Window.between(x.start, x.startInclusivity, cursor, Exclusive));
+      for (int i = 1; i < numberOfSubSpans - 1; i++) {
+        final var nextCursor = Duration.add(cursor, width);
+        ret.add(Window.between(cursor, Exclusive, nextCursor, Exclusive));
```

Comment:
Why are the intervals exclusive at both ends ? should it not be inclusive for start exclusive for end ? Shouldn't the union of all split intervals be equal to the original interval ?

Reply:
I know it is part of your specification to "remove" points but I am wondering why.

Reply:
For Windows, if any of the internal boundaries were inclusive, the Windows class would coalesce them back into the original window, so they have to be exclusive. The original external boundaries keep the same inclusivity though.

For Spans, we could have it that the spans meet at a point, but I want it to have the same general result as Windows so that if the user converts to windows, their split spans don't coalesce back.

Reply:
> If we talk about what it would mean in terms of scheduling, it would mean we can't schedule activities starting at those points.

I'd love it if we could have the sub-windows be left-inclusive and right-exclusive for exactly that reason, but unfortunately `[0, 1)    [1, 2)` is equal to `[0, 2)` because they contain the same points. So they are coalesced.

Reply:
I thought one of the motivation for adding `Spans` was that we could allow the correct behavior for splitting. Actually, the `split` operation, whatever the object it is called from, should return a `Spans` just because it inherently produces sets of contiguous intervals. I understand it is more beautiful if we have only one implementation and it is generic, etc, but if we do that, the meaning of the inclusivity bounds is kind of lost.
When the user goes from `Spans` to `Windows`, he knows/should know that contiguous intervals will be coalesced.

Reply:
Ok, I *almost* agree. I think it will be rare in scheduling to need to split windows. But what about constraints and global scheduling conditions, where the final object is Windows? It might be contrived, but I don't think its unreasonable to need it. Example:

> There must be at least one ActivityA every hour (i.e. locked to the hours of the clock, instead of a recurrence-goal-like constraint where we care about spacing)

I'd like to write:

```
dateResource.changes().invert()
    .split(24)
    .instanceCardinality(ActivityType.ActivityA)
    .greaterThan(0)
```

But if split produces Spans that will coalesce when converted to Windows, I'll have to write this:

```
dateResource.changes().invert()
    .split(24)
    .setBoundsInclusivity(Exclusive, Exclusive)
    .windows()
    .instanceCardinality(ActivityType.ActivityA) // this function won't be defined on Spans
    .greaterThan(0)
```

I guess that's not so bad. If you think that's a reasonable pattern, I'm OK making `split` always produce Spans.

Reply:
I thought everything related with defining goals was going to happen from `Spans` and not from `Windows` so I assume the first example you wrote would work.

Reply:
That first example is for a constraint, sorry that wasn't clear. This kind of use case would usually be covered by a goal, but perhaps they want to place the needed activities by hand. Again, kinda contrived, I know.

Reply:
> More generally, how should this behave on unbounded intervals?

I suppose as long as we're trying to represent infinite intervals on a finite domain, simply leaving windows that touch `MIN_VALUE` or `MAX_VALUE` unchanged is fairly accurate.

Reply:
I've resolved this with a new design: `split` always produces `Spans`. The user can specify what the inclusivity of new start/endpoints should be, and the default is [inclusive, exclusive). If the user wants windows instead of spans, they can convert it back with `.windows()` easily.